shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -393,30 +391,30 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
         val tn = td.nme
         val decls = MutMap.empty[Str, PolymorphicType]
         val defs = MutMap.empty[Str, PolymorphicType]
-        val targsMap = ctx.getTargsMaps(td.nme.name, N).getOrElse(Map.empty)
         val thisCtx = ctx.nest
-        val m = MutMap.empty[SimpleType, SimpleType]
-        thisCtx.targsMaps += td.nme.name -> MutMap(N -> targsMap.map { 
-          case s -> (tv: TypeVariable) =>
-            val rigid = freshenAbove(ctx.lvl, tv, true)
-            m += rigid -> tv
-            s -> rigid
+        val targsMap = td.targsMap.map { 
+          case s -> (tv: TypeVariable) => s -> freshenAbove(ctx.lvl, tv, true)
           case _ => die
-        })
-        val reverseRigid = m.toMap
-        thisCtx += "this" -> thisCtx.thisType(tn.name)
+        }
+        val rigid = td.targsMap.map { case pn -> tv => tv -> targsMap(pn) }.toMap[SimpleType, SimpleType]
+        val reverseRigid = rigid.map(_.swap).toMap
+        thisCtx += "this" -> subst(thisCtx.thisType(tn.name), rigid)
         (td.mthDecls ++ td.mthDefs).foreach { case md @ MethodDef(rec, prt, nme, tparams, rhs) =>
           implicit val prov: TypeProvenance = tp(md.toLoc, rhs.fold(_ => "method definition", _ => "method declaration"))
           val dummyTargs2 = tparams.map(p => freshVar(originProv(p.toLoc, "method type parameter"), S(p.name))(ctx.lvl + 2))
-          ctx.targsMaps(tn.name) += S(nme.name) -> tparams.map(_.name).zip(dummyTargs2).toMap
-          // ^ overwrites. fine for now as the maps are not used afterwards. may pass vars instead
+          val targsMap2 = tparams.map(_.name).zip(dummyTargs2).toMap
           if (!nme.name.head.isUpper)
             err(msg"Method names must start with a capital letter", nme.toLoc)
           if (defs.isDefinedAt(nme.name))
             err(msg"Method '${tn}.${nme.name}' is already defined.", nme.toLoc)
+          (targsMap.keySet & targsMap2.keySet).foreach { s =>
+            warn((s"Method type parameter ${targsMap2(s).nameHint.getOrElse("")}" -> targsMap2(s).prov.loco)
```

Comment:
This warning may be better reported by just looking at the syntactic class definition tree.
But in any case this logic is weird. You're doing unorthodox acrobatics with things like `targsMap2(s).nameHint.getOrElse("")`. Why not just `s`? If you stored Var instead of Str in targsMap you'd actually get a proper provenance as well as the name.